File: Java/task5.java
import java.util.ArrayList;
import java.util.Scanner;

public class Main {
    
    public static void generation(ArrayList<Integer> mod, int X0, int A, int B, int C, int[] repeat) {
        int Xi;
        boolean firstAppear = false;

        for (int i = 0; i < C; i++) {
            Xi = (A * mod.get(i) + B) % C;
            mod.add(Xi);
            if (X0 == Xi && mod.size() > 1 && !firstAppear) {
                repeat[0] = i + 2; // +2 потому что в индекс с нуля и прибавляется знач только на след итерации
                firstAppear = true;
            }
        }
    }

    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);
        ArrayList<Integer> mod = new ArrayList<>();
        int[] repeat = new int[1]; // Используем массив для передачи по ссылке
        int X0, A, B, C;

        System.out.println("Enter x0, A, B, C");
        X0 = scanner.nextInt();
        A = scanner.nextInt();
        B = scanner.nextInt();
        C = scanner.nextInt();

        if (!(0 <= A && A <= C) || !(0 <= B && B <= C) || !(0 <= X0 && X0 < C)) {
            System.out.println("error");
            scanner.close();
            return;
        }
        
        mod.add(X0);
        generation(mod, X0, A, B, C, repeat);

        for (int n : mod) {
            System.out.print(n + " ");
        }

        System.out.println();
        System.out.println(repeat[0]);

        scanner.close();
    }
}

File: C++/task5.cpp
#include <iostream>
#include <vector>
using namespace std;

void generation(vector <int>& mod, int X0, int A, int B, int C, int& repeat){

    int Xi; 
    bool firstAppear = false;   

    for (int i=0; i<C; i++){
        Xi = (A*mod[i] + B) % C;
        mod.push_back(Xi);
        if (X0==Xi && mod.size()>1 && firstAppear==false){
            repeat = i + 2; // +2 потому что в индекс с нуля и прибавляется знач только на след итерации
            firstAppear = true;
        }
    }
}

int main () {
    vector <int> mod;
    int repeat;
    int X0, A, B, C;

    cout << "Enter x0, A, B, C" << endl; 
    cin >> X0 >> A >> B >> C;

    if (!(0<=A && A<=C) || (!(0<=B && B<=C)) || (!(0<=X0 && X0 < C))){
        cout << "error";
        return -1;
    }
    mod.push_back(X0);

    generation(mod,X0, A, B, C, repeat);

    for (int n : mod){
        cout << n << " ";
    }

    cout << endl;
    cout << repeat;
}